File: Data Structures & Algorithms/intersection-of-two-linked-lists/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
"""
Understanding:
Provided two linkedlist headA and headB, return the intersecting node if they intersects.
Return None otherwise.

Planning:
Iterate through two linkedlist, check if the node shares the same memory address as we go
Starts from the node where we align their last node, so we can check if they are on the same node

   4->1->8->4->5
5->6->1->8->4->5
"""


class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        a = 0
        b = 0
        curr_a = headA
        curr_b = headB
        while curr_a:
            curr_a = curr_a.next
            a += 1

        while curr_b:
            curr_b = curr_b.next
            b += 1
        print(f"a={a}, b={b}")
        curr_a = headA
        curr_b = headB

        if a < b:
            for _ in range(b-a):
                curr_b = curr_b.next
        elif a > b:
            for _ in range(a-b):
                curr_a = curr_a.next
        
        k = min(a, b)

        for _ in range(k):
            if curr_a == curr_b:
                return curr_a

            curr_a = curr_a.next
            curr_b = curr_b.next

        return None

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
   public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        vector<int> nodes;

        for (int i = 0; i < static_cast<int>(lists.size()); ++i) {
            while (lists[i] != nullptr) {
                nodes.push_back((lists[i]->val));
                lists[i] = lists[i]->next;
            }
        }
        sort(nodes.begin(), nodes.end());

        ListNode* result = new ListNode(0);
        ListNode* curr = result;

        for (int i = 0; i < static_cast<int>(nodes.size()); ++i) {
            curr->next = new ListNode(nodes[i]);
            curr = curr->next;
        }
        return result->next;
    }
};
